tests/mock/clitools.rs

```diff
@@ -354,6 +354,22 @@ pub fn env(config: &Config, cmd: &mut Command) {
     cmd.env("RUSTUP_INIT_SKIP_PATH_CHECK", "yes");
 }
 
+use std::sync::RwLock;
+
+/// Returns the lock to be used when creating test environments.
+///
+/// Essentially we use this in `.read()` mode to gate access to `fork()`
+/// new subprocesses, and in `.write()` mode to gate creation of new test
+/// environments. In doing this we can ensure that new test environment creation
+/// does not result in ETXTBSY because the FDs in question happen to be in
+/// newly `fork()`d but not yet `exec()`d subprocesses of other tests.
```

Comment:
I don't understand why this is needed; copying files is a readonly - shared access ok - operation. Copying files that are in use as an executed binary should be ok as long as you open it appropriately.

Reply:
This is more that copying the binary involves a writable FD open for the target, and then if that FD is held open in a forked, but not yet exec'd child process of another test, then when the test which has now completed its copy attempts to exec that copy, the writable FD in the *other* test's forked subprocess prevents the exec.  There could be more precise locking semantics, but this doesn't degrade test performance too much and was the simplest I could come up with which didn't seem to have that race condition.